### PR TITLE
Add functions and objects for Temporary Directories

### LINKED ISSUE #232: An API for getting temp path in Filesystem.hh
## Desired behavior

`Filesystem.hh` should provide a function for getting a temp path suitable for the current operating system.

## Alternatives considered

N/A

## Implementation suggestion

Use the behavior as described in here:
https://www.boost.org/doc/libs/1_75_0/libs/filesystem/doc/reference.html#temp_directory_path
To cover both Linux/macOS and Windows.

## Additional context

`common::cwd()` is often used to "get a path to save a file" currently.
The problem with this approach is that CWD is not necessarily writable.
`av/src/VideoEncoder.cc` contains an example of this pattern.

### PR DESCRIPTION
**Pull request by mjcarroll**
# 🎉 Functions for interacting with temporary directories

Closes #232.  Split from #238 to make a version that should be portable to all ign-common4 platforms.

## Summary

Adds free functions and a RAII object for interacting with operating system temporary directories.  This is useful for writing tests or utilities that need to access a temporary location that can be cleaned up at the end of execution, rather than using build artifact directories or users' home folders.

Free Functions:
* `tempDirectoryPath()` - retrieve OS-specific location of temporary directory
* `createTempDirectory()` - create a directory in the `tempDirectoryPath`, expanding template strings where necessary.

Obejcts:
* `TempDirectory` - Object that will create a temporary directory and optionally clean it up when it goes out of scope.

## Test it

Usage is demonstrated in the `TempDirectory_TEST.cc` unit tests.

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

File: src/TempDirectory.cc [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        << " (ec: " << _ec << " " << _ec.message() << ")\n";
+    }
+    return false;
+  }
+  return true;
+}
+
+
+/////////////////////////////////////////////////
+std::string temp_directory_path(std::error_code& _err)
```
Comment (chapulina) [inline src/TempDirectory.cc:52]: Style nit: 
```suggestion
std::string tempDirectoryPath(std::error_code& _err)
```

There are also instances within the function

File: src/TempDirectory.cc [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+/// was copied from rclcpp.
+std::string createTempDirectory(
+    const std::string &_baseName,
+    const std::string &_parentPath)
+{
+  std::string parentPath(_parentPath);
+  std::string templatePath = _baseName + "XXXXXX";
+
+  std::string fullTemplateStr = joinPaths(parentPath, templatePath);
+  if (!createDirectories(parentPath)) {
```
Comment (chapulina) [inline src/TempDirectory.cc:100]:
```suggestion
  if (!createDirectories(parentPath))
  {
```

File: src/TempDirectory.cc [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+/////////////////////////////////////////////////
+std::string ignition::common::createTempDirectory(
+    const std::string &_baseName,
+    const std::string &_parentPath,
+    const FilesystemWarningOp _warningOp)
+{
+  std::string ret;
+  try {
+    ret = ::createTempDirectory(_baseName, _parentPath);
+  } catch (const std::system_error &ex) {
```
Comment (chapulina) [inline src/TempDirectory.cc:142]:
```suggestion
  }
  catch (const std::system_error &_ex)
  {
```

File: include/ignition/common/TempDirectory.hh [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+      ///
+      /// This is useful if you wish to clean by default during a test, but
+      /// retain the contents of the TempDirectory if the test fails.
+      /// \param[in] _cleanup True to indicate that the filesystem should
+      ///  be cleaned as part of the destructor
+      public: void Cleanup(bool _cleanup);
+
+      /// \brief Retrieve the current cleanup flag state
+      /// \return true if filesystem cleanup will occur
+      public: bool Cleanup() const;
```
Comment (chapulina) [inline include/ignition/common/TempDirectory.hh:93]: These functions' names could be misleading, sounding like the cleanup happens when they're called

File: src/TempDirectory.cc [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+  auto tempPath = common::tempDirectoryPath();
+  if (!_subDir.empty())
+  {
+    tempPath = common::joinPaths(tempPath, _subDir);
+  }
+  this->dataPtr->path = common::createTempDirectory(_prefix, tempPath);
+  if (!this->dataPtr->path.empty())
+  {
+    this->dataPtr->isValid = true;
+    common::chdir(this->dataPtr->path);
```
Comment (chapulina) [inline src/TempDirectory.cc:187]: Can you document that the working directory will be changed at construction and destruction?

File: src/TempDirectory_TEST.cc [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    EXPECT_TRUE(tmp.Cleanup());
+    path = tmp.Path();
+    EXPECT_FALSE(path.empty());
+    EXPECT_TRUE(ignition::common::exists(path));
+  }
+  EXPECT_FALSE(ignition::common::exists(path));
+}
+
+/////////////////////////////////////////////////
+TEST(TempDirectory, TempDirectory_noclean)
```
Comment (chapulina) [inline src/TempDirectory_TEST.cc:84]:
```suggestion
TEST(TempDirectory, TempDirectoryNoClean)
```

File: src/TempDirectory_TEST.cc [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    path = tmp.Path();
+    EXPECT_FALSE(path.empty());
+    EXPECT_TRUE(ignition::common::exists(path));
+  }
+  EXPECT_TRUE(ignition::common::exists(path));
+  EXPECT_TRUE(ignition::common::removeDirectory(path));
+}
+
+/////////////////////////////////////////////////
+TEST(TempDirectory, TempDirectory_noclean_later)
```
Comment (chapulina) [inline src/TempDirectory_TEST.cc:100]:
```suggestion
TEST(TempDirectory, TempDirectoryNoCleanLater)
```

**Review by chapulina — APPROVED** [review]
Nice! I just have minor comments, but LGTM.

File: src/TempDirectory.cc (hunk 1 above)
Comment (mjcarroll) [reply src/TempDirectory.cc:52]: This is a local implementation of std::filesystem::temp_directory_path.  It is not part of the public API and can be removed as soon as std::filesystem is supported everywhere necessary.  I have noted this in a comment.

File: include/ignition/common/TempDirectory.hh (hunk 4 above)
Comment (mjcarroll) [reply include/ignition/common/TempDirectory.hh:93]: I changed them to `DoCleanup`, which mirrors what the internal member is also called.

File: src/TempDirectory.cc (hunk 5 above)
Comment (mjcarroll) [reply src/TempDirectory.cc:193]: ee7cae2

File: include/ignition/common/TempDirectory.hh [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    /// Upon construction, the current working directory will be set to this
+    /// new temporary directory.
+    /// Upon destruction, the current working directory will be restored to the
+    /// location when the TempDirectory object was constructed.
+    class IGNITION_COMMON_VISIBLE TempDirectory
+    {
+      /// \brief Create a directory in the tempDirectoryPath by expanding
+      /// a name template.  This directory can also be automatically cleaned
+      /// up when the object goes out of scope.
+      ///
```
Comment (ahcorde) [inline include/ignition/common/TempDirectory.hh:67]:
```suggestion
      /// a name template. This directory can also be automatically cleaned
      /// up when the object goes out of scope.
      ///
```

File: include/ignition/common/TempDirectory.hh [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+      /// \brief Indicate if the TempDirectory object is in a valid state
+      /// and that the folder exists on the filesystem
+      /// \return true if the TempDirectory is valid
+      public: bool Valid() const;
+
+      /// \brief Set if the folder on disk should be cleaned.
+      ///
+      /// This is useful if you wish to clean by default during a test, but
+      /// retain the contents of the TempDirectory if the test fails.
+      /// \param[in] _cleanup True to indicate that the filesystem should
```
Comment (ahcorde) [inline include/ignition/common/TempDirectory.hh:91]:
```suggestion
      /// \param[in] _doCleanup True to indicate that the filesystem should
```

File: src/TempDirectory.cc [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+using namespace ignition;
+using namespace common;
+
+/////////////////////////////////////////////////
+// Return true if success, false if error
+inline bool fs_warn(const std::string &_fcn,
+             const std::error_code &_ec,
+             const FilesystemWarningOp &_warningOp = FSWO_LOG_WARNINGS)
+{
+
```
Comment (ahcorde) [inline src/TempDirectory.cc:37]:
```suggestion
```

File: src/TempDirectory_TEST.cc [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <gtest/gtest.h>
+
+#include <ignition/utilities/ExtraTestMacros.hh>
+#include <ignition/common/TempDirectory.hh>
```
Comment (ahcorde) [inline src/TempDirectory_TEST.cc:21]: alphabetize

File: src/TempDirectory.cc [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+#include <ignition/common/TempDirectory.hh>
+#include <ignition/common/Console.hh>
+
+#ifdef _WIN32
+#include <windows.h>
+#include <direct.h>
+#include <fileapi.h>
+#include <io.h>
+#endif
```
Comment (ahcorde) [inline src/TempDirectory.cc:26]: alphabetize

File: include/ignition/common/TempDirectory.hh [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+      /// \brief Retrieve the current cleanup flag state
+      /// \return true if filesystem cleanup will occur
+      public: bool DoCleanup() const;
+
+      /// \brief Retrieve the fully-expanded temporary directory path
+      /// \return the temporary directory path
+      public: std::string Path() const;
+
+      IGN_COMMON_WARN_IGNORE__DLL_INTERFACE_MISSING
+      private: std::unique_ptr<TempDirectoryPrivate> dataPtr;
```
Comment (ahcorde) [inline include/ignition/common/TempDirectory.hh:104]: `IGN_UTILS_IMPL_PTR` ?

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: include/ignition/common/TempDirectory.hh (hunk 13 above)
Comment (mjcarroll) [reply include/ignition/common/TempDirectory.hh]: 3fd9e27

File: src/TempDirectory_TEST.cc (hunk 11 above)
Comment (mjcarroll) [reply src/TempDirectory_TEST.cc]: 3fd9e27

File: src/TempDirectory.cc [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+    if (FSWO_LOG_WARNINGS == _warningOp)
+    {
+      ignwarn << "Failed ignition::common::" << _fcn
+        << " (ec: " << _ec << " " << _ec.message() << ")\n";
+    }
+    return false;
+  }
+  return true;
+}
+
```
Comment (ahcorde) [inline src/TempDirectory.cc:49]:
```suggestion
```

**Review by ahcorde — APPROVED** [review]

**Comment by mjcarroll** [comment]
Failure in focal CI is due to flaky test that uses timings.

**Review by ahcorde — DISMISSED** [review]
Windows is failing https://build.osrfoundation.org/job/ignition_common-ci-pr_any-windows7-amd64/1390/consoleFull#15419766088ff58640-3599-4406-a210-216932f1748c

**Comment by mjcarroll** [comment]
> Windows is failing https://build.osrfoundation.org/job/ignition_common-ci-pr_any-windows7-amd64/1390/consoleFull#15419766088ff58640-3599-4406-a210-216932f1748c

This will require switching ignition common to use `colcon`:  https://github.com/ignition-tooling/release-tools/pull/511

**Comment by mjcarroll** [comment]
@osrf-jenkins retest this please

**Comment by mjcarroll** [comment]
@ahcorde should be good now.  `windows.h` has to be first, which breaks when alphabetized.

**Review by chapulina — APPROVED** [review]

File: src/TempDirectory.cc (hunk 1 above)
Comment (mjcarroll) [reply src/TempDirectory.cc:54]: Here is where it is removed in the switch to std::filesystem:

https://github.com/ignitionrobotics/ign-common/pull/238/files#diff-cc8a2d7ec160eade949b51ef4324d5784f463d271eea0c3fd23b52e14e057a74L50-L74

File: src/CMakeLists.txt [hunk 15]
```diff
@@ -17,6 +17,9 @@ target_link_libraries(${PROJECT_LIBRARY_TARGET_NAME}
 target_include_directories(${PROJECT_LIBRARY_TARGET_NAME} PRIVATE
   ${ignition-math${IGN_MATH_VER}_INCLUDE_DIRS})
 
+target_include_directories(${PROJECT_LIBRARY_TARGET_NAME} PUBLIC
+  ${ignition-utils${IGN_UTILS_VER}_INCLUDE_DIRS})
```
Comment (scpeters) [inline src/CMakeLists.txt:21]: this means ign-utils1 is now required by the core `ignition-common4` library. we should update the find_package call to make this dependency `REQUIRED`

**Comment by scpeters** [comment]
this pull request may not work properly on debian buster with armhf: https://github.com/ignitionrobotics/ign-common/issues/255

**Pull request merged**